### PR TITLE
More Standard Library Improvements

### LINKED ISSUE #1279: `File.open` method
### Summary
While our `File` library provides users with good levels of control, it is somewhat lacking in terms of ease of use. We want to add a `File.open -> Text -> Text` that opens and reads the file in one shot.

### Value
Files can be opened and read in one shot.

### Specification

- [x] Add a `File.open` method that opens and reads the file in one go.

### Acceptance Criteria & Test Cases

- [x] Files can be opened and read in one function call.

### PR DESCRIPTION
**Pull request by iamrecursion**
### Pull Request Description
Fixes a bunch of stuff and adds new features.

Closes #1279.

### Important Notes
N/A

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/main/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/main/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: distribution/std-lib/Base/src/Data/Any/Extensions.enso [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+Any.is_nothing = case this of
+    Nothing -> True
+    _ -> False
+
+## Checks if the type is not an instance of `Nothing`.
+
+   > Example
+     Checking if a variable `a` is not nothing.
+         a.not_nothing
+Any.not_nothing : Boolean
```
Comment (kustosz) [inline distribution/std-lib/Base/src/Data/Any/Extensions.enso]: Remove this method – too much methods on `Any` and a clumsy name.

File: distribution/std-lib/Base/src/Data/Map.enso [hunk 2]
```diff
@@ -4,6 +4,32 @@ import Base.Data.Map.Internal
 ## An error for getting a missing value from a map.
 type No_Value_For_Key key
 
+## Returns an empty map.
+
+   > Example
+     Create an empty map.
```
Comment (kustosz) [inline distribution/std-lib/Base/src/Data/Map.enso]: Useless example – it's exactly the same as just pressing enter in the searcher

File: distribution/std-lib/Base/src/Data/Map.enso [hunk 3]
```diff
@@ -4,6 +4,32 @@ import Base.Data.Map.Internal
...
+
+## Builds a map from a vector of key-value pairs.
+
+   > Example
+     Building a map containing two key-value pairs.
+         Map.from_vector [[1, 2], [3, 4]]
+from_vector : Vector.Vector -> Map
+from_vector vec =
+    stop = vec.length
+    go map index = if index >= stop then map else
```
Comment (kustosz) [inline distribution/std-lib/Base/src/Data/Map.enso]: Unnecessary recursion, use `vec.fold` instead.

File: distribution/std-lib/Base/src/Data/Map.enso [hunk 4]
```diff
@@ -16,6 +42,10 @@ type Map
         Bin _ _ _ _ _ -> False
         Tip -> True
 
+    ## Checks if the map is not empty.
+    not_empty : Boolean
+    not_empty = not this.is_empty
```
Comment (kustosz) [inline distribution/std-lib/Base/src/Data/Map.enso]: write it as `this.is_empty.not` – less work when we migrate to non-UFCS

File: distribution/std-lib/Base/src/Data/Map.enso [hunk 5]
```diff
@@ -49,34 +79,167 @@ type Map
...
+    ## Inserts a key-value mapping into this map. If `key` is already present,
+       it will be overriden with the new `value`.
+    insert : Any -> Any -> Map
+    insert key value = Internal.insert this key value
 
     ## Gets the value associated with `key` in this map, or returns a
-       `No_Value_For_Key` error, if `key` is not present.
-    get : Any -> Any ! No_Value_For_Key
+       `Nothing`, if `key` is not present.
+    get : Any -> Any ! Nothing
```
Comment (kustosz) [inline distribution/std-lib/Base/src/Data/Map.enso]:
```suggestion
    get : Any -> Any | Nothing
```

File: distribution/std-lib/Base/src/Data/Map.enso [hunk 6]
```diff
@@ -49,34 +79,167 @@ type Map
...
+    ## Apply a function to each value in the map.
+
+       This method does not return the results, so is only useful for performing
+       computations with side-effects.
+
+       > Example
+         Printing each value in the map `m`.
+             m.each IO.println
+    each : (Any -> Any) -> Nothing
+    each function =
```
Comment (kustosz) [inline distribution/std-lib/Base/src/Data/Map.enso:150]: Write this from scratch – you're allocating a copy of this map just to drop it.

File: distribution/std-lib/Base/src/Data/Map.enso [hunk 7]
```diff
@@ -49,34 +79,167 @@ type Map
...
+
+       > Example
+         Printing each key and value in the map `m`.
+             m.each (k -> v -> IO.println (k.to_text + v.to_text))
+    each_with_key : (Any -> Any -> Any) -> Nothing
+    each_with_key function =
+        this.map_with_key function
+        Nothing
+
+    ## Fold over the values in the map.
```
Comment (kustosz) [inline distribution/std-lib/Base/src/Data/Map.enso]: `Fold over the values` means nothing to a non-functional non-dev. Reword with something like "combine all values..."

File: distribution/std-lib/Base/src/Data/Map.enso [hunk 8]
```diff
@@ -49,34 +79,167 @@ type Map
...
+    fold init function =
+        go map init = case map of
+            Bin _ _ v l r ->
+                y = go l init
+                z = function y v
+                go r z
+            Tip -> init
+        go this init
+
+    ## Fold over the key-value pairs in the map.
```
Comment (kustosz) [inline distribution/std-lib/Base/src/Data/Map.enso]: Same comment as `fold`

File: distribution/std-lib/Base/src/Data/Range.enso [hunk 9]
```diff
@@ -86,7 +86,11 @@ type Range
          Checking that at least one number in the range is greater than 10.
              1.up_to 100 . exists (> 10)
     exists : (Number -> Boolean) -> Boolean
-    exists predicate = this.fold False (l -> r -> l || predicate r)
+    exists predicate = #this.fold False (l -> r -> l || predicate r)
```
Comment (kustosz) [inline distribution/std-lib/Base/src/Data/Range.enso]: stray comment

File: distribution/std-lib/Base/src/Data/Text/Extensions.enso [hunk 10]
```diff
@@ -109,28 +118,36 @@ Text.< that = Text_Utils.lt [this, that]
 Text.> : Text -> Boolean
 Text.> that = Text_Utils.lt [that, this]
 
+## Check if `this` is empty.
+Text.is_empty : Boolean
+Text.is_empty = this == ""
+
+## Check if `this` is not empty.
+Text.not_empty : Boolean
```
Comment (kustosz) [inline distribution/std-lib/Base/src/Data/Text/Extensions.enso:126]: `non_empty` seems nicer to me. Your call tho :)

File: distribution/std-lib/Base/src/Data/Time/Locale.enso [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+taiwan : Locale
+taiwan = here.from_java (Polyglot.get_member JavaLocale "TAIWAN")
+
+## The locale for the UK.
+uk : Locale
+uk = here.from_java (Polyglot.get_member JavaLocale "UK")
+
+## The locale for the US.
+us : Locale
+us = here.from_java (Polyglot.get_member JavaLocale "US")
```
Comment (kustosz) [inline distribution/std-lib/Base/src/Data/Time/Locale.enso]: What's the key to this assortment of countries? Seems pretty strange to me. Misses some large countries (India, Russia, Brazil) while exposing, e.g. Taiwan. I'd go by countries-with-highest-population (or countries with highest population + countries with highest GDP, or some other metric like that)

File: distribution/std-lib/Base/src/Data/Time/Locale.enso [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    language : Text | Nothing
+    language =
+        lang = this.java_locale.getLanguage []
+        if lang.is_empty then Nothing else lang
+
+    ## Get the country from the locale.
+    country : Text | Nothing
+    country =
+        place = this.java_locale.getCountry []
+        if place.is_empty then Nothing else lang
```
Comment (kustosz) [inline distribution/std-lib/Base/src/Data/Time/Locale.enso]: forgot to change `lang` in these copy-pastes

**Review by kustosz — APPROVED** [review]
Very cool additions! There are some problems that must be addressed before merge, but I'm approving because they don't require a re-review.

**Review by 4e6 — APPROVED** [review]
LGTM

**Review by kustosz — APPROVED** [review]
Looks good, one stray comment left :)

**Pull request merged**